src/System.Text.Json/tests/JsonReaderTests.cs

```diff
@@ -0,0 +1,160 @@
+            Assert.Equal(ExpectedId2, Convert.ToInt32(actualIDs[1]));
+            Assert.Equal(ExpectedId3, Convert.ToInt64(actualIDs[2]));
+        }
+
+        private const string Json =
+            "{\"age\":30,\"balance\":-1000,\"first\":\"John\",\"last\":\"Smith\",\"phoneNumbers\":[\"425-000-1212\",\"425-000-1213\"],\"address\":{\"street\":\"1 Microsoft Way\",\"city\":\"Redmond\",\"zip\":98052}\"IDs\" \n : [ \t 0425 \n , -70 \n, 9223372036854775807 ] \t }";
+
+        private const uint ExpectedAge = 30;
+        private const int ExpectedBalance = -1000;
+        private const uint ExpectedId1 = 425;
+        private const int ExpectedId2 = -70;
+        private const long ExpectedId3 = 9223372036854775807;
+        private readonly Utf8String _expectedFullName = new Utf8String("John Smith");
+
+        private static readonly Utf8String[] ExpectedPhone =
+        {
+            new Utf8String("425-000-1212"),
+            new Utf8String("425-000-1213")
+        };
+
+        private readonly Utf8String _expectedAddress = new Utf8String("1 Microsoft Way, Redmond, 98052");
+
+        private static List<KeyValuePair<Utf8String, object>> Read(JsonReader json)
+        {
+            var jsonOutput = new List<KeyValuePair<Utf8String, object>>();
+            var property = new Utf8String("");
+            while (json.Read())
+            {
+                object value;
+                switch (json.TokenType)
+                {
+                    case JsonReader.JsonTokenType.ObjectStart:
+                        json.ReadObjectStart();
+                        value = null;
+                        break;
+                    case JsonReader.JsonTokenType.ObjectEnd:
+                        json.ReadObjectEnd();
+                        property = new Utf8String("");
+                        value = null;
+                        break;
+                    case JsonReader.JsonTokenType.ArrayStart:
+                        json.ReadArrayStart();
+                        value = null;
+                        break;
+                    case JsonReader.JsonTokenType.ArrayEnd:
+                        json.ReadArrayEnd();
+                        property = new Utf8String("");
+                        value = null;
+                        break;
+                    case JsonReader.JsonTokenType.PropertyName:
+                        property = json.ReadPropertyAsString();
+                        value = null;
+                        break;
+                    case JsonReader.JsonTokenType.PropertyValueAsString:
+                        value = json.ReadPropertyAsString();
+                        break;
+                    case JsonReader.JsonTokenType.PropertyValueAsInt:
+                        value = json.ReadPropertyValueAsInt();
+                        break;
+                    default:
+                        property = new Utf8String("");
+                        value = null;
+                        break;
```

Comment:
I would simplify some of these more, but for now, let's check in and then talk about additional changes.